File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean isValid(int[] map1, int[] map2) {
        for(int i = 0; i < 26; i++) if(map1[i] != map2[i]) return false;
        return true;
    }
    public boolean checkInclusion(String s1, String s2) {
        int n = s1.length(), m = s2.length(), l = 0;
        int[] map1 = new int[26];
        int[] map2 = new int[26];

        if(n > m) return false;

        for(int r = 0; r < m; r++){
            if(r < n) {
                char ch1 = s1.charAt(r);
                map1[ch1 - 'a']++;                
            }
            
            char ch2 = s2.charAt(r);
            map2[ch2 - 'a']++;
            
            if(r >= n - 1) {
                if(isValid(map1, map2)) return true;
                char toRemove = s2.charAt(l++);
                map2[toRemove - 'a']--;   
            }
        }
        return false;
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if(s1.size() > s2.size()) return false;
        vector<int> mp1 (26, 0);
        vector<int> mp2 (26, 0);
        int l = 0, r = 0, len = s1.size();
        while(r < len){
            mp1[s1[r] - 'a']++;
            mp2[s2[r] - 'a']++;
            r++;
        }
        if(mp1 == mp2) return true;
        while(r < s2.size()){
            mp2[s2[r] - 'a']++;
            mp2[s2[l] - 'a']--;
            if(mp1 == mp2) return true;
            r++; l++;
        }
        return false;
    }
};